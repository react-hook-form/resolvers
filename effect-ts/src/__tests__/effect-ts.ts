import { Schema } from 'effect';
import { effectTsResolver } from '..';
import { fields, invalidData, schema, validData } from './__fixtures__/data';

const shouldUseNativeValidation = false;

describe('effectTsResolver', () => {
  it('should return values from effectTsResolver when validation pass', async () => {
    const result = await effectTsResolver(schema)(validData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toEqual({ errors: {}, values: validData });
  });

  it('should return a single error from effectTsResolver when validation fails', async () => {
    const result = await effectTsResolver(schema)(invalidData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toMatchSnapshot();
  });

  it('should return the first error from effectTsResolver when validation fails with `validateAllFieldCriteria` set to firstError', async () => {
    const SignupSchema = Schema.Struct({
      phoneNumber: Schema.optional(
        Schema.String.pipe(
          Schema.pattern(/^\+\d{7,15}$/, {
            message: () =>
              'Please enter a valid phone number in international format.',
          }),
        ),
      ),
    });

    const result = await effectTsResolver(SignupSchema)(
      { phoneNumber: '123' },
      undefined,
      {
        fields: {
          phoneNumber: {
            ref: { name: 'phoneNumber' },
            name: 'phoneNumber',
          },
        },
        criteriaMode: 'firstError',
        shouldUseNativeValidation,
      },
    );

    expect(result).toMatchSnapshot();
  });

  it('should return all the errors from effectTsResolver when validation fails with `validateAllFieldCriteria` set to true', async () => {
    const SignupSchema = Schema.Struct({
      phoneNumber: Schema.optional(
        Schema.String.pipe(
          Schema.pattern(/^\+\d{7,15}$/, {
            message: () =>
              'Please enter a valid phone number in international format.',
          }),
        ),
      ),
    });

    const result = await effectTsResolver(SignupSchema)(
      { phoneNumber: '123' },
      undefined,
      {
        fields: {
          phoneNumber: {
            ref: { name: 'phoneNumber' },
            name: 'phoneNumber',
          },
        },
        criteriaMode: 'all',
        shouldUseNativeValidation,
      },
    );

    expect(result).toMatchSnapshot();
  });
});
