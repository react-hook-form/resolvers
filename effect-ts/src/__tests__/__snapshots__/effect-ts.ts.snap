// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`effectTsResolver > should return a single error from effectTsResolver when validation fails 1`] = `
{
  "errors": {
    "animal": {
      "message": "Expected string, actual ["dog"]",
      "ref": undefined,
      "type": "Type",
    },
    "luckyNumbers": [
      ,
      ,
      {
        "message": "Expected number, actual "3"",
        "ref": undefined,
        "type": "Type",
      },
    ],
    "password": {
      "message": "At least 1 special character.",
      "ref": {
        "name": "password",
      },
      "type": "Refinement",
    },
    "vehicles": [
      ,
      {
        "horsepower": {
          "message": "is missing",
          "ref": undefined,
          "type": "Missing",
        },
      },
    ],
  },
  "values": {},
}
`;

exports[`effectTsResolver > should return all the errors from effectTsResolver when validation fails with \`validateAllFieldCriteria\` set to true 1`] = `
{
  "errors": {
    "phoneNumber": {
      "message": "Please enter a valid phone number in international format.",
      "ref": {
        "name": "phoneNumber",
      },
      "type": "Refinement",
      "types": {
        "Refinement": "Please enter a valid phone number in international format.",
        "Type": "Expected undefined, actual "123"",
      },
    },
  },
  "values": {},
}
`;

exports[`effectTsResolver > should return the first error from effectTsResolver when validation fails with \`validateAllFieldCriteria\` set to firstError 1`] = `
{
  "errors": {
    "phoneNumber": {
      "message": "Please enter a valid phone number in international format.",
      "ref": {
        "name": "phoneNumber",
      },
      "type": "Refinement",
    },
  },
  "values": {},
}
`;
